sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCheckZeroSizeCollection.cs

```diff
@@ -0,0 +1,144 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotCheckZeroSizeCollection : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3981";
+        private const string MessageFormat = "The {0} of '{1}' is always '>=0', so fix this test to get the real expected behavior.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var gteNode = (BinaryExpressionSyntax)c.Node;
+                    CheckCondition(c, gteNode.Left, gteNode.Right);
+                }, SyntaxKind.GreaterThanOrEqualExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var lteNode = (BinaryExpressionSyntax)c.Node;
+                    CheckCondition(c, lteNode.Right, lteNode.Left);
+                }, SyntaxKind.LessThanOrEqualExpression);
+        }
+
+        private static void CheckCondition(SyntaxNodeAnalysisContext context, ExpressionSyntax expressionValueNode,
+            ExpressionSyntax constValueNode)
+        {
+            if (!IsConstantZero(context, constValueNode))
+            {
+                return;
+            }
+
+            ISymbol symbol = GetSymbol(context, expressionValueNode);
+
+            string symbolType = null;
```

Comment:
I would extract this line and the following if/else-if statements into `string GetDeclaringTypeName(ISymbol)`